### PR TITLE
Address performance of smart indenter at end of file with inactive regions

### PR DESCRIPTION
**Pull request by pharring**
Customer reported performance issue navigating through a file with lots of inactive regions.

### CODE REVIEW

**Comment by pharring** [comment]
Note: I tried to make the same fix for VB, but it is missing the concept of ".IsActive" for directive syntax. I'll enter a separate issue for that.

**Comment by pharring** [comment]
@basoundr, @DustinCampbell for review.

**Comment by DustinCampbell** [comment]
:+1:

**Comment by pharring** [comment]
This change breaks some keyword recommender tests around "elif". Investigating.

File: src/Workspaces/CSharp/Portable/Extensions/SyntaxNodeExtensions.cs [hunk 1]
```diff
@@ -780,10 +780,29 @@ public static IEnumerable<SyntaxNode> GetAncestorsOrThis(this SyntaxNode node, F
         /// </summary>
         private static IEnumerable<SyntaxToken> GetSkippedTokens(SyntaxTriviaList list)
         {
+            // PERF: Avoid allocations in the most common case of no skipped tokens.
+            if (!HasSkippedTokens(list))
```
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/Extensions/SyntaxNodeExtensions.cs]: Is enumerating the list twice really better than the allocation?  What about just unwrapping the query below?

File: src/Workspaces/CSharp/Portable/Extensions/SyntaxTreeExtensions.cs [hunk 2]
```diff
@@ -534,24 +534,26 @@ private static bool AtEndOfIncompleteStringOrCharLiteral(SyntaxToken token, int
...
             if (token.Kind() == SyntaxKind.EndOfFileToken)
             {
+                var text = syntaxTree.GetText(cancellationToken);
+                var lineContainingPosition = text.Lines.IndexOf(position);
                 var triviaList = token.LeadingTrivia;
                 foreach (var triviaTok in triviaList.Reverse())
                 {
-                    if (triviaTok.HasStructure)
+                    var triviaLine = text.Lines.IndexOf(triviaTok.SpanStart);
+                    if (triviaLine <= lineContainingPosition)
```
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/Extensions/SyntaxTreeExtensions.cs]: Deliberate change from `<` to `<=`?

**Comment by Pilchie** [comment]
:+1: apart from the question on `<` vs `<=`.

File: src/Workspaces/CSharp/Portable/Extensions/SyntaxTreeExtensions.cs [hunk 3]
```diff
@@ -534,24 +534,28 @@ private static bool AtEndOfIncompleteStringOrCharLiteral(SyntaxToken token, int
...
             var token = syntaxTree.FindTokenOrEndToken(position, cancellationToken);
-            var text = syntaxTree.GetText(cancellationToken);
-            var lineContainingPosition = text.Lines.IndexOf(position);
             if (token.Kind() == SyntaxKind.EndOfFileToken)
             {
                 var triviaList = token.LeadingTrivia;
                 foreach (var triviaTok in triviaList.Reverse())
                 {
-                    if (triviaTok.HasStructure)
+                    if (triviaTok.Span.Contains(position))
```
Comment (DustinCampbell) [inline src/Workspaces/CSharp/Portable/Extensions/SyntaxTreeExtensions.cs]: Ah. Oops. :smile:

**Pull request merged**